beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -3077,62 +3187,105 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-                    .eth1_finalization_cache
-                    .try_write_for(ETH1_FINALIZATION_CACHE_LOCK_TIMEOUT)
-                    .and_then(|mut cache| {
-                        cache.insert(checkpoint, eth1_finalization_data);
-                        cache.finalize(&current_finalized_checkpoint)
-                    })
-                {
-                    if let Some(eth1_chain) = self.eth1_chain.as_ref() {
-                        let finalized_deposit_count = finalized_eth1_data.deposit_count;
-                        eth1_chain.finalize_eth1_data(finalized_eth1_data);
-                        debug!(
-                            self.log,
-                            "called eth1_chain.finalize_eth1_data()";
-                            "epoch" => current_finalized_checkpoint.epoch,
-                            "deposit count" => finalized_deposit_count,
-                        );
-                    }
-                }
+    fn import_block_update_shuffling_cache(
+        &self,
+        block_root: Hash256,
+        state: &mut BeaconState<T::EthSpec>,
+    ) -> Result<(), BlockError<T::EthSpec>> {
+        // For the current and next epoch of this state, ensure we have the shuffling from this
+        // block in our cache.
+        for relative_epoch in [RelativeEpoch::Current, RelativeEpoch::Next] {
+            let shuffling_id = AttestationShufflingId::new(block_root, state, relative_epoch)?;
+
+            let shuffling_is_cached = self
+                .shuffling_cache
+                .try_read_for(ATTESTATION_CACHE_LOCK_TIMEOUT)
+                .ok_or(Error::AttestationCacheLockTimeout)?
+                .contains(&shuffling_id);
+
+            if !shuffling_is_cached {
+                state.build_committee_cache(relative_epoch, &self.spec)?;
+                let committee_cache = state.committee_cache(relative_epoch)?;
+                self.shuffling_cache
+                    .try_write_for(ATTESTATION_CACHE_LOCK_TIMEOUT)
+                    .ok_or(Error::AttestationCacheLockTimeout)?
+                    .insert_committee_cache(shuffling_id, committee_cache);
             }
         }
+        Ok(())
+    }
 
-        // Inform the unknown block cache, in case it was waiting on this block.
-        self.pre_finalization_block_cache
-            .block_processed(block_root);
+    #[allow(clippy::too_many_arguments)]
+    fn import_block_update_deposit_contract_finalization(
+        &self,
+        block: BeaconBlockRef<T::EthSpec>,
+        block_root: Hash256,
+        current_epoch: Epoch,
+        current_finalized_checkpoint: Checkpoint,
+        current_eth1_finalization_data: Eth1FinalizationData,
+        parent_eth1_finalization_data: Eth1FinalizationData,
+        parent_block_slot: Slot,
+    ) {
+        // Do not write to eth1 finalization cache for blocks older than 5 epochs.
+        if block.slot().epoch(T::EthSpec::slots_per_epoch()) + 5 < current_epoch {
```

Comment:
I changed this condition slightly to simplify the inputs to this function. Would appreciate you double-checking it @ethDreamer :pray:

Reply:
@michaelsproul LGTM! Very clean :)

Reply:
Thanks!